core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java

```diff
@@ -77,10 +77,11 @@
     private Duration interruptStuckSplitTasksDetectionInterval = new Duration(2, TimeUnit.MINUTES);
 
     private boolean scaleWritersEnabled = true;
-    // The default value is 8 because it is better in performance compare to 2 or 4
-    // and acceptable in terms of resource utilization since values like 32 or higher could take
-    // more resources, hence potentially affect the other concurrent queries in the cluster.
-    private int scaleWritersMaxWriterCount = 8;
+    // Set the value of default max writer count to the number of processors and cap it to 32. We can do this
+    // because preferred write partitioning is always enabled for local exchange thus partitioned inserts will never
+    // use this property. Hence, there is no risk in terms of more numbers of physical writers which can cause high
+    // resource utilization.
+    private int scaleWritersMaxWriterCount = min(nextPowerOfTwo(getAvailablePhysicalProcessorCount()), 32);
```

Comment:
Note that getAvailablePhysicalProcessorCount could give different number on coordinator and worker. Would that cause any problems ? It looks like we use this value in `PipelinedQueryScheduler` on coordinator and `LocalExecutionPlanner` on worker.